Update spec for MSC3952: intentional mentions (#1508)
matrix-org/matrix-spec: content/client-server-api/modules/mentions.md

### User and room mentions

> {{< changed-in v="1.7" >}}

Reviewer: sorry, I think this should have been %s to use the right spacing (for reasons I have never quite taken the time to grok)

```suggestion
{{% changed-in v="1.7" %}}
```

Reviewer: similarly for the push rules

This module allows users to "mention" other users and rooms within a room event.
This is primarily used as an indicator that the recipient should receive a notification
about the event.
This is achieved by including metadata in the `m.mentions` content property of
the event to reference the entity being mentioned.

`m.mentions` is defined as follows:

{{% definition path="api/client-server/definitions/m.mentions" %}}

Additionally, see the [`.m.rule.is_user_mention`](#_m_rule_is_user_mention) and
[`.m.rule.is_room_mention`](#_m_rule_is_room_mention) push rules.
Users should not add their own Matrix ID to the `m.mentions` property as outgoing
messages cannot self-notify.

Note that, for backwards compatibility, push rules such as [`.m.rule.contains_display_name`](#_m_rule_contains_display_name),
[`.m.rule.contains_user_name`](#_m_rule_contains_user_name), and
[`.m.rule.roomnotif`](#_m_rule_roomnotif) continue to  match if the `body` of
the event contains the user's display name or ID. To avoid unintentional notifications,
**it is recommended that clients include a `m.mentions` property on each event**.
(If there are no mentions to include it can be an empty object.)

> {% boxes/rationale %}

Reviewer: ```suggestion
{{% boxes/rationale %}}
```

In previous versions of the specification, mentioning users was done by
including the user's display name or the localpart of their Matrix ID and room
mentions were done by including the string "@room" in the plaintext `body` of
the event. This was prone to confusing and buggy behaviour.

> {% /boxes/rationale %}

Reviewer: ```suggestion
{{% /boxes/rationale %}}
```

> ### User and room mentions

Author: I would be curious if this is the right spot to document `m.mentions`?

Reviewer: I'd have thought so.

### User and room mentions

> 

Reviewer: ```suggestion

{{< changed-in v="1.7" >}}

```

This module allows users to mention other users and rooms within a room message.
This is achieved by including metadata in the `m.mentions` content property of

> the event. The `m.mentions` property consists of an object with two fields:

Reviewer: I wonder if we need to make it explicit that `m.mentions` is encrypted in the same way as the rest of `content`?

Author: I assumed that any property not given as an exception is encrypted by default?

Reviewer: I hesitate because (a) there was debate about it in the MSC stage, and people might still have the wrong impression; (b) afaik the only other "mixin" property we have like this is `m.relates_to`, which is *not* encrypted.

I think a couple of words making it explicit wouldn't hurt.

Author: That's fair, I added a note about ti and pointed back to how clients should process push rules (for encrypted events).

[Matrix URI](/appendices/#uris) in the HTML body of  an [m.room.message](#mroommessage)
event. This applies only to [m.room.message](#mroommessage) events where the `msgtype` is
`m.text`, `m.emote`, or `m.notice`. The `format` for the event must be
`org.matrix.custom.html` and therefore requires a `formatted_body`.

To make a mention, reference the entity being mentioned in the
`m.mentions` property, like so:

```json
{
    "body": "Hello Alice!",
    "msgtype": "m.text",
    "format": "org.matrix.custom.html",
    "formatted_body": "Hello <a href='https://matrix.to/#/@alice:example.org'>Alice</a>!",
    "m.mentions": {
        "user_ids": ["@alice:example.org"]
    }
}
```

Similarly, the entire room can be mentioned:

```json
{
    "body": "Hello @room!",
    "msgtype": "m.text",
    "format": "org.matrix.custom.html",
    "formatted_body": "Hello @room!",
    "m.mentions": {
        "room": true
    }
}
```

Additionally, see [the `.m.rule.is_user_mention` and `.m.rule.is_room_mention` push rules](#predefined-rules).
Users should not add their own Matrix ID to the `m.mentions` property as outgoing
messages cannot self-notify.

> 

Reviewer: It would be nice to include some text in here about how this changed in this spec version.

Normally we do that with a `{% boxes/rationale %}` box (for some reason)

> To disable legacy behaviour of notifications occurring due to matching against
> the localpart of Matrix IDs and display names it is recommended that clients include
> a `m.mentions` property on each event. If there are no mentions to include it can
> be an empty object.

Reviewer: This is a bit hard to grok for a first-time reader who is unaware of this legacy behaviour.

```suggestion
Note that, for backwards compatibility, push rules such as `.m.rule.contains_user_name` <link> continue to
match if the *content* of the event contains the user's name or ID. To avoid unintentional notifications,
**it is recommended that clients include a `m.mentions` property on each event**. (If there are no mentions to include it can be an empty object.)
```

#### Client behaviour

> In addition to using the appropriate `Matrix URI` for the mention,

Reviewer: I'm not convinced the text in this section is still accurate. In particular, it seems to treat "mention" as equivalent to "matrix URI link", which now seems incorrect.

Author: I think this section is mostly about *displaying events with mentions in them*? So yes, it needs updating but I'm not quite sure what to do there. Will take an attempt.

Author: I think most of it is accurate-ish, I did attempt to clean it up a bit though. WDYT?

Reviewer: I think it's much better now, thanks!

example, this may be done by changing the background color of the
mention to indicate that it is different from a normal link.

If the current user is mentioned in a message, the client should show that
mention differently from other mentions, such as by using a red
background color to signify to the user that they were mentioned.

When clicked, the mention should navigate the user to the appropriate
user or room information.

TODO Behavior with replies.

> TODO Behavior with edits.

Author: Does it make sense to include information here about handling replies / edits or in their respective sections? I think here (and then cross-reference), but I'm not sure.

Reviewer: My instinct would actually be to go the other way around (keep the mentions section simple, and put the thorny details of the interactions with replies and edits in their respective sections). There is already an [edits of replies](https://spec.matrix.org/v1.6/client-server-api/#edits-of-replies) section, and by analogy I think an "edits of mentions" section would make sense.

Not a strong opinion though.

Author: Edits of replies which include mentions does get thorny! 👍 That sounds like a better place indeed! Thank you!

example, this may be done by changing the background color of the
mention to indicate that it is different from a normal link.

If the current user is mentioned in a message, the client should show that
mention differently from other mentions, such as by using a red

> background color to signify to the user that they were mentioned.

Reviewer: there is text below here that says group mentions look the same as room mentions. That is certainly no longer true. I feel like we should just remove the stuff about groups.
